{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4f70f5a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gffpandas.gffpandas as gffpd\n",
    "\n",
    "import pandas as pd\n",
    "from pandas.core.common import SettingWithCopyWarning\n",
    "import gseapy as gp\n",
    "import warnings\n",
    "\n",
    "import numpy as np\n",
    "from gseapy.plot import barplot, dotplot\n",
    "from matplotlib import pyplot as plt \n",
    "\n",
    "\n",
    "warnings.simplefilter(action=\"ignore\", category=SettingWithCopyWarning)\n",
    "\n",
    "path_to_data='/Users/aleksandradenisova/Desktop/coursework/data/stage2'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b4f168df",
   "metadata": {},
   "outputs": [],
   "source": [
    "annotation = gffpd.read_gff3('{}/{}'.format(path_to_data, 'GCF_003957565.2_bTaeGut1.4.pri_genomic.gff'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "6c6803a5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>chrom</th>\n",
       "      <th>GeneId</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_044211.2</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>gene</td>\n",
       "      <td>6695</td>\n",
       "      <td>12279</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene-LOC121470777;Dbxref=GeneID:121470777;N...</td>\n",
       "      <td>NC_044211.2</td>\n",
       "      <td>LOC121470777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_044211.2</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>lnc_RNA</td>\n",
       "      <td>6695</td>\n",
       "      <td>12279</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna-XR_005981978.1;Parent=gene-LOC121470777...</td>\n",
       "      <td>NC_044211.2</td>\n",
       "      <td>XR_005981978.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_044211.2</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>6695</td>\n",
       "      <td>6759</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon-XR_005981978.1-1;Parent=rna-XR_0059819...</td>\n",
       "      <td>NC_044211.2</td>\n",
       "      <td>XR_005981978.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_044211.2</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>7790</td>\n",
       "      <td>7996</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon-XR_005981978.1-2;Parent=rna-XR_0059819...</td>\n",
       "      <td>NC_044211.2</td>\n",
       "      <td>XR_005981978.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NC_044211.2</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>8177</td>\n",
       "      <td>12279</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon-XR_005981978.1-3;Parent=rna-XR_0059819...</td>\n",
       "      <td>NC_044211.2</td>\n",
       "      <td>XR_005981978.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        seq_id  source     type  start    end score strand phase  \\\n",
       "1  NC_044211.2  Gnomon     gene   6695  12279     .      +     .   \n",
       "2  NC_044211.2  Gnomon  lnc_RNA   6695  12279     .      +     .   \n",
       "3  NC_044211.2  Gnomon     exon   6695   6759     .      +     .   \n",
       "4  NC_044211.2  Gnomon     exon   7790   7996     .      +     .   \n",
       "5  NC_044211.2  Gnomon     exon   8177  12279     .      +     .   \n",
       "\n",
       "                                          attributes        chrom  \\\n",
       "1  ID=gene-LOC121470777;Dbxref=GeneID:121470777;N...  NC_044211.2   \n",
       "2  ID=rna-XR_005981978.1;Parent=gene-LOC121470777...  NC_044211.2   \n",
       "3  ID=exon-XR_005981978.1-1;Parent=rna-XR_0059819...  NC_044211.2   \n",
       "4  ID=exon-XR_005981978.1-2;Parent=rna-XR_0059819...  NC_044211.2   \n",
       "5  ID=exon-XR_005981978.1-3;Parent=rna-XR_0059819...  NC_044211.2   \n",
       "\n",
       "           GeneId  \n",
       "1    LOC121470777  \n",
       "2  XR_005981978.1  \n",
       "3  XR_005981978.1  \n",
       "4  XR_005981978.1  \n",
       "5  XR_005981978.1  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ori = annotation.df\n",
    "df = df_ori[df_ori['type'].isin(['gene', 'lnc_RNA', 'exon', 'mRNA', 'CDS', 'tRNA',\n",
    "       'transcript', 'primary_transcript', 'miRNA', 'pseudogene',\n",
    "       'snoRNA', 'guide_RNA', 'snRNA', 'C_gene_segment', 'V_gene_segment', 'rRNA', 'sequence_feature'])]\n",
    "\n",
    "df['GeneId'] = df['attributes'].apply(lambda x: x.split(';')[0].split('=')[1].split('-')[1])\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "34699397",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#GeneName</th>\n",
       "      <th>GeneId</th>\n",
       "      <th>TranscriptId</th>\n",
       "      <th>BioType</th>\n",
       "      <th>variants_impact_LOW</th>\n",
       "      <th>variants_impact_MODERATE</th>\n",
       "      <th>variants_impact_MODIFIER</th>\n",
       "      <th>variants_effect_3_prime_UTR_variant</th>\n",
       "      <th>variants_effect_5_prime_UTR_premature_start_codon_gain_variant</th>\n",
       "      <th>variants_effect_5_prime_UTR_variant</th>\n",
       "      <th>variants_effect_downstream_gene_variant</th>\n",
       "      <th>variants_effect_intron_variant</th>\n",
       "      <th>variants_effect_missense_variant</th>\n",
       "      <th>variants_effect_non_coding_transcript_exon_variant</th>\n",
       "      <th>variants_effect_splice_region_variant</th>\n",
       "      <th>variants_effect_synonymous_variant</th>\n",
       "      <th>variants_effect_upstream_gene_variant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ADAMTS2</td>\n",
       "      <td>ADAMTS2</td>\n",
       "      <td>XM_030283925.3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ADSS2</td>\n",
       "      <td>ADSS2</td>\n",
       "      <td>XM_012572420.4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ADSS2</td>\n",
       "      <td>ADSS2</td>\n",
       "      <td>XM_041715289.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AFAP1L2</td>\n",
       "      <td>AFAP1L2</td>\n",
       "      <td>XM_030276462.3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AFAP1L2</td>\n",
       "      <td>AFAP1L2</td>\n",
       "      <td>XM_030276463.3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  #GeneName   GeneId    TranscriptId         BioType  variants_impact_LOW  \\\n",
       "0   ADAMTS2  ADAMTS2  XM_030283925.3  protein_coding                    0   \n",
       "1     ADSS2    ADSS2  XM_012572420.4  protein_coding                    0   \n",
       "2     ADSS2    ADSS2  XM_041715289.1  protein_coding                    0   \n",
       "3   AFAP1L2  AFAP1L2  XM_030276462.3  protein_coding                    0   \n",
       "4   AFAP1L2  AFAP1L2  XM_030276463.3  protein_coding                    0   \n",
       "\n",
       "   variants_impact_MODERATE  variants_impact_MODIFIER  \\\n",
       "0                         0                         2   \n",
       "1                         0                         3   \n",
       "2                         0                         3   \n",
       "3                         0                         2   \n",
       "4                         0                         2   \n",
       "\n",
       "   variants_effect_3_prime_UTR_variant  \\\n",
       "0                                    0   \n",
       "1                                    0   \n",
       "2                                    0   \n",
       "3                                    0   \n",
       "4                                    0   \n",
       "\n",
       "   variants_effect_5_prime_UTR_premature_start_codon_gain_variant  \\\n",
       "0                                                  0                \n",
       "1                                                  0                \n",
       "2                                                  0                \n",
       "3                                                  0                \n",
       "4                                                  0                \n",
       "\n",
       "   variants_effect_5_prime_UTR_variant  \\\n",
       "0                                    0   \n",
       "1                                    0   \n",
       "2                                    0   \n",
       "3                                    0   \n",
       "4                                    0   \n",
       "\n",
       "   variants_effect_downstream_gene_variant  variants_effect_intron_variant  \\\n",
       "0                                        0                               2   \n",
       "1                                        0                               3   \n",
       "2                                        0                               3   \n",
       "3                                        0                               2   \n",
       "4                                        0                               2   \n",
       "\n",
       "   variants_effect_missense_variant  \\\n",
       "0                                 0   \n",
       "1                                 0   \n",
       "2                                 0   \n",
       "3                                 0   \n",
       "4                                 0   \n",
       "\n",
       "   variants_effect_non_coding_transcript_exon_variant  \\\n",
       "0                                                  0    \n",
       "1                                                  0    \n",
       "2                                                  0    \n",
       "3                                                  0    \n",
       "4                                                  0    \n",
       "\n",
       "   variants_effect_splice_region_variant  variants_effect_synonymous_variant  \\\n",
       "0                                      0                                   0   \n",
       "1                                      0                                   0   \n",
       "2                                      0                                   0   \n",
       "3                                      0                                   0   \n",
       "4                                      0                                   0   \n",
       "\n",
       "   variants_effect_upstream_gene_variant  \n",
       "0                                      0  \n",
       "1                                      0  \n",
       "2                                      0  \n",
       "3                                      0  \n",
       "4                                      0  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_genes = pd.read_csv('~/Desktop/coursework/data/final_data/positions_0.05_ANOVA_inno_zibra.vcf.txt',\n",
    "            sep='\\t',\n",
    "            skiprows=1)\n",
    "df_genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "d90e6571",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['gene', nan], dtype=object)"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_mereged = df_genes.merge(df[['GeneId','type']], on = 'GeneId', how='left')\n",
    "df_mereged['type'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "5216e0a8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#GeneName</th>\n",
       "      <th>GeneId</th>\n",
       "      <th>TranscriptId</th>\n",
       "      <th>BioType</th>\n",
       "      <th>variants_impact_LOW</th>\n",
       "      <th>variants_impact_MODERATE</th>\n",
       "      <th>variants_impact_MODIFIER</th>\n",
       "      <th>variants_effect_3_prime_UTR_variant</th>\n",
       "      <th>variants_effect_5_prime_UTR_premature_start_codon_gain_variant</th>\n",
       "      <th>variants_effect_5_prime_UTR_variant</th>\n",
       "      <th>variants_effect_downstream_gene_variant</th>\n",
       "      <th>variants_effect_intron_variant</th>\n",
       "      <th>variants_effect_missense_variant</th>\n",
       "      <th>variants_effect_non_coding_transcript_exon_variant</th>\n",
       "      <th>variants_effect_splice_region_variant</th>\n",
       "      <th>variants_effect_synonymous_variant</th>\n",
       "      <th>variants_effect_upstream_gene_variant</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1884</th>\n",
       "      <td>TRNAP-AGG_1</td>\n",
       "      <td>TRNAP-AGG_1</td>\n",
       "      <td>unassigned_transcript_15</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1885</th>\n",
       "      <td>TRNAP-CGG</td>\n",
       "      <td>TRNAP-CGG</td>\n",
       "      <td>unassigned_transcript_14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        #GeneName       GeneId              TranscriptId         BioType  \\\n",
       "1884  TRNAP-AGG_1  TRNAP-AGG_1  unassigned_transcript_15  protein_coding   \n",
       "1885    TRNAP-CGG    TRNAP-CGG  unassigned_transcript_14  protein_coding   \n",
       "\n",
       "      variants_impact_LOW  variants_impact_MODERATE  variants_impact_MODIFIER  \\\n",
       "1884                    0                         0                         1   \n",
       "1885                    0                         0                         1   \n",
       "\n",
       "      variants_effect_3_prime_UTR_variant  \\\n",
       "1884                                    0   \n",
       "1885                                    0   \n",
       "\n",
       "      variants_effect_5_prime_UTR_premature_start_codon_gain_variant  \\\n",
       "1884                                                  0                \n",
       "1885                                                  0                \n",
       "\n",
       "      variants_effect_5_prime_UTR_variant  \\\n",
       "1884                                    0   \n",
       "1885                                    0   \n",
       "\n",
       "      variants_effect_downstream_gene_variant  variants_effect_intron_variant  \\\n",
       "1884                                        1                               0   \n",
       "1885                                        0                               0   \n",
       "\n",
       "      variants_effect_missense_variant  \\\n",
       "1884                                 0   \n",
       "1885                                 0   \n",
       "\n",
       "      variants_effect_non_coding_transcript_exon_variant  \\\n",
       "1884                                                  0    \n",
       "1885                                                  0    \n",
       "\n",
       "      variants_effect_splice_region_variant  \\\n",
       "1884                                      0   \n",
       "1885                                      0   \n",
       "\n",
       "      variants_effect_synonymous_variant  \\\n",
       "1884                                   0   \n",
       "1885                                   0   \n",
       "\n",
       "      variants_effect_upstream_gene_variant type  \n",
       "1884                                      0  NaN  \n",
       "1885                                      1  NaN  "
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_mereged[df_mereged['type'].isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "2df3dd0c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>smth</th>\n",
       "      <th>INFO</th>\n",
       "      <th>type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>effect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_044211.2</td>\n",
       "      <td>123676</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>T,C,A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ANN=A|intron_variant|MODIFIER|CMSS1|CMSS1|tran...</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>CMSS1</td>\n",
       "      <td>MODIFIER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_044211.2</td>\n",
       "      <td>544931</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>T,C,A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ANN=A|intron_variant|MODIFIER|LOC100222962|LOC...</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>LOC100222962</td>\n",
       "      <td>MODIFIER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_044211.2</td>\n",
       "      <td>2727694</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G,C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ANN=C|intron_variant|MODIFIER|MYH15|MYH15|tran...</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>MYH15</td>\n",
       "      <td>MODIFIER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_044211.2</td>\n",
       "      <td>2760307</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>T,G,C,A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ANN=A|intron_variant|MODIFIER|CIP2A|CIP2A|tran...</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>CIP2A</td>\n",
       "      <td>MODIFIER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_044211.2</td>\n",
       "      <td>5048551</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ANN=A|intron_variant|MODIFIER|LOC100220100|LOC...</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>LOC100220100</td>\n",
       "      <td>MODIFIER</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        #CHROM      POS ID REF      ALT FILTER smth  \\\n",
       "0  NC_044211.2   123676  .   T    T,C,A      .    .   \n",
       "1  NC_044211.2   544931  .   T    T,C,A      .    .   \n",
       "2  NC_044211.2  2727694  .   A      G,C      .    .   \n",
       "3  NC_044211.2  2760307  .   T  T,G,C,A      .    .   \n",
       "4  NC_044211.2  5048551  .   G        A      .    .   \n",
       "\n",
       "                                                INFO            type  \\\n",
       "0  ANN=A|intron_variant|MODIFIER|CMSS1|CMSS1|tran...  intron_variant   \n",
       "1  ANN=A|intron_variant|MODIFIER|LOC100222962|LOC...  intron_variant   \n",
       "2  ANN=C|intron_variant|MODIFIER|MYH15|MYH15|tran...  intron_variant   \n",
       "3  ANN=A|intron_variant|MODIFIER|CIP2A|CIP2A|tran...  intron_variant   \n",
       "4  ANN=A|intron_variant|MODIFIER|LOC100220100|LOC...  intron_variant   \n",
       "\n",
       "      gene_name    effect  \n",
       "0         CMSS1  MODIFIER  \n",
       "1  LOC100222962  MODIFIER  \n",
       "2         MYH15  MODIFIER  \n",
       "3         CIP2A  MODIFIER  \n",
       "4  LOC100220100  MODIFIER  "
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ann = pd.read_csv('/Users/aleksandradenisova/programs/snpEff/positions_0.05_ANOVA_inno_zibra.vcf.ann', \n",
    "                     skiprows=6,\n",
    "                     sep = '\\t', \n",
    "                     names = ['#CHROM', 'POS', 'ID', 'REF', 'ALT', 'FILTER', 'smth','INFO'])\n",
    "\n",
    "df_ann['type'] = df_ann['INFO'].apply(lambda x: x.split('|')[1])\n",
    "df_ann['gene_name'] = df_ann['INFO'].apply(lambda x: x.split('|')[3])\n",
    "df_ann['effect'] = df_ann['INFO'].apply(lambda x: x.split('|')[2])\n",
    "df_ann.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "67156cea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>smth</th>\n",
       "      <th>INFO</th>\n",
       "      <th>type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>effect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NC_044211.2</td>\n",
       "      <td>14926206</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ANN=C|upstream_gene_variant|MODIFIER|TRNAP-CGG...</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>TRNAP-CGG</td>\n",
       "      <td>MODIFIER</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        #CHROM       POS ID REF ALT FILTER smth  \\\n",
       "7  NC_044211.2  14926206  .   T   C      .    .   \n",
       "\n",
       "                                                INFO                   type  \\\n",
       "7  ANN=C|upstream_gene_variant|MODIFIER|TRNAP-CGG...  upstream_gene_variant   \n",
       "\n",
       "   gene_name    effect  \n",
       "7  TRNAP-CGG  MODIFIER  "
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ann[df_ann['gene_name']=='TRNAP-CGG']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d288a8f",
   "metadata": {},
   "source": [
    "WARNING_TRANSCRIPT_NO_START_CODON: игнорируем, нашей тРНК старт-кодон и не нужен"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "93eba9b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ANN=C|upstream_gene_variant|MODIFIER|TRNAP-CGG|TRNAP-CGG|transcript|unassigned_transcript_14|protein_coding||c.-3282T>C|||||3282|WARNING_TRANSCRIPT_NO_START_CODON',\n",
       " 'C|downstream_gene_variant|MODIFIER|TRNAP-AGG_1|TRNAP-AGG_1|transcript|unassigned_transcript_15|protein_coding||c.*3748A>G|||||3748|WARNING_TRANSCRIPT_NO_START_CODON',\n",
       " 'C|intron_variant|MODIFIER|RCSD1|RCSD1|transcript|XM_030263664.3|protein_coding|6/6|c.*13+386T>C||||||',\n",
       " 'C|intron_variant|MODIFIER|RCSD1|RCSD1|transcript|XM_030263670.3|protein_coding|5/5|c.*13+386T>C||||||']"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ann[df_ann['gene_name']=='TRNAP-CGG']['INFO'].loc[7].split(',')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d373640a",
   "metadata": {},
   "source": [
    "#### Вывод: \n",
    "\n",
    "есть предположение, что эта мутация находится на участке между пролиновыми тРНК: после последовательности, которая кодирует тРНК с антикодоном AGG, но перед последовательностью, кодирующей тРНК с антикодоном CGG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1930c1c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "vscode": {
   "interpreter": {
    "hash": "b29209ab99999b3e97b72049df2199dd7f57ed342fe759e68413d592d2f571a7"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
